File: crates/draco-cpp-test-bridge/cpp/encode_drc.cpp
// Encodes an .obj N times and nothing else -- the C++ half of the encode
// callgrind comparison, matching examples/encode_drc.rs step for step.
//
// Built directly against a Draco checkout rather than through the Rust
// bridge, because the comparison runs where valgrind does (Linux/WSL) and the
// bridge links a Windows library. See PERFORMANCE.md, the encode callgrind
// round, for the build line.
//
// Both sides read the same .obj, each through its own parser, so the meshes
// are only equal if the encoders agree on the output. The byte count printed
// here is that check: compare it against the Rust driver's before reading a
// single per-stage figure. The parse itself is outside the loop on both
// sides and is not part of what is being compared.
//
//   g++ -O2 -DNDEBUG -std=c++17 -I<src> -I<build> encode_drc.cpp \
//       -L<build> -ldraco -o encode_drc
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <iterator>
#include <vector>

#include "draco/core/decoder_buffer.h"

#include "draco/compression/encode.h"
#include "draco/io/obj_decoder.h"
#include "draco/mesh/mesh.h"

int main(int argc, char** argv) {
    if (argc < 3) {
        std::fprintf(stderr, "usage: encode_drc <file.obj> <speed> [iters]\n");
        return 2;
    }
    const int speed = std::atoi(argv[2]);
    const int iters = argc > 3 ? std::atoi(argv[3]) : 1;

    // Read the file here rather than through `DecodeFromFile`: a statically
    // linked Draco has no file reader registered -- the factory's registration
    // lives in a translation unit the linker drops -- so that path fails with
    // "Unable to read input file" on a perfectly good `.obj`.
    std::ifstream file(argv[1], std::ios::binary);
    if (!file) {
        std::fprintf(stderr, "cannot open %s\n", argv[1]);
        return 2;
    }
    std::vector<char> obj_text((std::istreambuf_iterator<char>(file)),
                               std::istreambuf_iterator<char>());
    draco::DecoderBuffer obj_buffer;
    obj_buffer.Init(obj_text.data(), obj_text.size());

    draco::Mesh mesh;
    draco::ObjDecoder obj_decoder;
    const auto parsed = obj_decoder.DecodeFromBuffer(&obj_buffer, &mesh);
    if (!parsed.ok()) {
        std::fprintf(stderr, "cannot parse %s: %s\n", argv[1], parsed.error_msg());
        return 2;
    }

    // The same options `common::options_for` builds: position at 11 bits, a
    // second attribute at 8, speed on both dials.
    draco::Encoder encoder;
    encoder.SetSpeedOptions(speed, speed);
    encoder.SetAttributeQuantization(draco::GeometryAttribute::POSITION, 11);
    if (mesh.num_attributes() > 1) {
        encoder.SetAttributeQuantization(draco::GeometryAttribute::NORMAL, 8);
        encoder.SetAttributeQuantization(draco::GeometryAttribute::TEX_COORD, 8);
    }

    size_t bytes = 0;
    for (int i = 0; i < iters; ++i) {
        draco::EncoderBuffer buffer;
        const auto status = encoder.EncodeMeshToBuffer(mesh, &buffer);
        if (!status.ok()) {
            std::fprintf(stderr, "encode failed: %s\n", status.error_msg());
            return 1;
        }
        bytes = buffer.size();
    }
    std::fprintf(stderr,
                 "cpp encoded %s at speed %d: %zu bytes from %d points / %d faces x%d\n",
                 argv[1], speed, bytes, (int)mesh.num_points(),
                 (int)mesh.num_faces(), iters);
    return 0;
}

File: crates/draco-cpp-test-bridge/examples/encode_drc.rs
//! Encodes an `.obj` N times and nothing else.
//!
//! The counterpart of `cpp/encode_drc.cpp`, and the encode half of the
//! callgrind comparison -- the same shape as `decode_drc`, for the same
//! reason: no bridge, no harness, so it builds anywhere `draco-core` does,
//! including the WSL side where `valgrind` lives.
//!
//! Both sides read the same `.obj` rather than a `.drc`, because encode has
//! no encoded input to share. What makes the comparison honest is the byte
//! count each side prints: two encoders that agree on the output size to the
//! byte were given the same mesh under the same options. Check it before
//! reading any per-stage figure.
//!
//! Under callgrind use `1` iteration: the tool counts instructions, so
//! repetition buys nothing and costs minutes.
//!
//! ```text
//! cargo run --release --example encode_drc -- seeded_grid_0.obj 5 1
//! valgrind --tool=callgrind --callgrind-out-file=rust.out ./encode_drc grid.obj 5 1
//! ```
use draco_core::encoder_buffer::EncoderBuffer;
use draco_core::mesh_encoder::MeshEncoder;

#[path = "common/mod.rs"]
mod common;
use common::{load, options_for};

fn main() {
    let mut args = std::env::args().skip(1);
    let path = args.next().expect("mesh path");
    let speed: i32 = args
        .next()
        .and_then(|s| s.parse().ok())
        .expect("encode speed");
    let iters: u32 = args.next().and_then(|s| s.parse().ok()).unwrap_or(1);

    let payload = load(&path);
    let options = options_for(&payload.mesh, speed);

    // The mesh is handed over once, outside the loop. `set_mesh` takes
    // ownership, so a per-iteration clone would charge this side a copy the
    // C++ entry point does not make -- and `encode_matrix` brackets it the
    // same way, outside its timer. `encode` resets its derived state, so the
    // encoder is reusable across iterations exactly as `draco::Encoder` is.
    let mut encoder = MeshEncoder::new();
    encoder.set_mesh(payload.mesh.clone());

    let mut bytes = 0;
    for _ in 0..iters {
        let mut buffer = EncoderBuffer::new();
        encoder
            .encode(&options, &mut buffer)
            .expect("rust encode failed");
        bytes = buffer.data().len();
    }
    eprintln!(
        "rust encoded {path} at speed {speed}: {bytes} bytes from {} points / {} faces x{iters}",
        payload.mesh.num_points(),
        payload.mesh.num_faces()
    );
}
